File: react-native-cliqit/ios/CliqItModule.swift
import CliqIt
import Foundation
import React

@objc(CliqItModule)
class CliqItModule: RCTEventEmitter {
  private var hasListeners = false
  // ponytail: RN listeners attach after configure; buffer one shot so events aren't dropped
  private var pendingDeepLink: [String: Any]?
  private var pendingLinkLookup: [String: Any]?
  private var pendingVerify: [String: Any]?

  override static func requiresMainQueueSetup() -> Bool { true }

  override func supportedEvents() -> [String]! {
    ["CliqItLinkReceived", "CliqItLinkLookup", "CliqItVerify"]
  }

  override func startObserving() {
    hasListeners = true
    flushPending()
  }

  override func stopObserving() {
    hasListeners = false
  }

  @objc
  func configure(_ apiKey: String) {
    CliqItSDK.shared.onVerify { [weak self] result in
      self?.emitVerify(result)
    }

    CliqItSDK.shared.configure(apiKey: apiKey)

    CliqItSDK.shared.onLinkReceived { [weak self] payload in
      self?.emitLinkReceived(payload)
    }

    CliqItSDK.shared.onDirectLinkLookup { [weak self] result in
      self?.emitLinkLookup(result)
    }

    CliqItSDK.shared.notifyAlreadyReportedIfNeeded()
  }

  @objc
  func handleUrl(_ urlString: String) {
    guard let url = URL(string: urlString) else { return }
    _ = CliqItSDK.shared.handle(url: url)
  }

  override func constantsToExport() -> [AnyHashable: Any]! {
    [
      "LinkField": [
        "destination": CliqItLinkField.destination.rawValue,
        "iosDestination": CliqItLinkField.iosDestination.rawValue,
        "androidDestination": CliqItLinkField.androidDestination.rawValue,
        "ogTitle": CliqItLinkField.ogTitle.rawValue,
        "ogDescription": CliqItLinkField.ogDescription.rawValue,
        "ogImage": CliqItLinkField.ogImage.rawValue,
        "ogUrl": CliqItLinkField.ogUrl.rawValue,
        "slug": CliqItLinkField.slug.rawValue,
        "webFallback": CliqItLinkField.webFallback.rawValue,
        "showInterstitial": CliqItLinkField.showInterstitial.rawValue,
        "isDeepLink": CliqItLinkField.isDeepLink.rawValue,
        "appleTeamId": CliqItLinkField.appleTeamId.rawValue,
        "iosBundleId": CliqItLinkField.iosBundleId.rawValue,
        "androidPackageName": CliqItLinkField.androidPackageName.rawValue,
        "resolvedPath": CliqItLinkField.resolvedPath.rawValue,
      ],
    ]
  }

  // MARK: - Emit / buffer

  private func emitLinkReceived(_ payload: CliqItPayload) {
    var body: [String: Any] = [
      "url": payload.url.absoluteString,
      "path": payload.path,
      "pathComponents": payload.pathComponents,
      "query": payload.queryParameters,
      "source": payload.source.rawValue,
      "isDeferred": payload.isDeferred,
      "status": payload.status.rawValue,
      "shouldNavigate": payload.shouldNavigate,
    ]
    if let matched = payload.matched { body["matched"] = matched }
    if let tier = payload.tier { body["tier"] = tier }
    if let confidence = payload.confidence { body["confidence"] = confidence }
    if let score = payload.score { body["score"] = score }
    if let slug = payload.slug { body["slug"] = slug }
    if let destinationPath = payload.destinationPath { body["destinationPath"] = destinationPath }
    if let errorMessage = payload.errorMessage { body["error"] = errorMessage }
    emitBody(event: "CliqItLinkReceived", body: body, pendingKey: \.pendingDeepLink)
  }

  private func emitLinkLookup(_ result: Result<CliqItLinkDetails, CliqItDeferredMatchError>) {
    let body: [String: Any]
    switch result {
    case .success(let details):
      body = [
        "status": "resolved",
        "destination": details[.destination] as Any,
        "iosDestination": details[.iosDestination] as Any,
        "androidDestination": details[.androidDestination] as Any,
        "ogTitle": details[.ogTitle] as Any,
        "ogDescription": details[.ogDescription] as Any,
        "ogImage": details[.ogImage] as Any,
        "ogUrl": details[.ogUrl] as Any,
        "slug": details[.slug] as Any,
        "webFallback": details[.webFallback] as Any,
        "showInterstitial": details[.showInterstitial] as Any,
        "isDeepLink": details[.isDeepLink] as Any,
        "appleTeamId": details[.appleTeamId] as Any,
        "iosBundleId": details[.iosBundleId] as Any,
        "androidPackageName": details[.androidPackageName] as Any,
        "resolvedPath": details[.resolvedPath] as Any,
      ]
    case .failure(let error):
      body = [
        "status": "failed",
        "error": error.localizedDescription,
      ]
    }
    emitBody(event: "CliqItLinkLookup", body: body, pendingKey: \.pendingLinkLookup)
  }

  private func emitVerify(_ outcome: CliqItVerifyOutcome) {
    var body: [String: Any] = [:]
    switch outcome {
    case .passed(let r):
      body = verifyBody(status: "ok", result: r)
    case .mismatched(let r):
      body = verifyBody(status: "mismatch", result: r)
      body["message"] = r.mismatchMessage
    case .error(let error):
      body = [
        "status": "failed",
        "ok": false,
        "error": error.localizedDescription,
      ]
    }
    if let raw = CliqItSDK.shared.currentVerifyJSON {
      body["raw"] = raw
    }
    emitBody(event: "CliqItVerify", body: body, pendingKey: \.pendingVerify)
  }

  private func verifyBody(status: String, result: CliqItVerifyResult) -> [String: Any] {
    var checks: [String: Any] = [:]
    for (key, check) in result.checks {
      checks[key] = [
        "actual": check.actual as Any,
        "expected": check.expected as Any,
        "match": check.match,
      ]
    }
    return [
      "status": status,
      "ok": result.ok,
      "appId": result.appId as Any,
      "appName": result.appName as Any,
      "checks": checks,
      "message": result.mismatchMessage,
    ]
  }

  private func emitBody(
    event: String,
    body: [String: Any],
    pendingKey: ReferenceWritableKeyPath<CliqItModule, [String: Any]?>
  ) {
    if hasListeners {
      sendEvent(withName: event, body: body)
    } else {
      self[keyPath: pendingKey] = body
    }
  }

  private func flushPending() {
    if let deep = pendingDeepLink {
      pendingDeepLink = nil
      sendEvent(withName: "CliqItLinkReceived", body: deep)
    }
    if let lookup = pendingLinkLookup {
      pendingLinkLookup = nil
      sendEvent(withName: "CliqItLinkLookup", body: lookup)
    }
    if let verify = pendingVerify {
      pendingVerify = nil
      sendEvent(withName: "CliqItVerify", body: verify)
    }
  }
}

File: react-native-cliqit/android/src/main/java/com/cliqit/CliqItModule.kt
package com.cliqit

import com.cliqit.sdk.CliqItSDK
import com.cliqit.sdk.LinkField
import com.cliqit.sdk.LinkLookupResult
import com.cliqit.sdk.VerifyOutcome
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.WritableMap
import com.facebook.react.modules.core.DeviceEventManagerModule

class CliqItModule(private val reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  private var hasListeners = false
  private var pendingDeepLink: WritableMap? = null
  private var pendingLinkLookup: WritableMap? = null
  private var pendingVerify: WritableMap? = null
  private var handlersBound = false

  override fun getName(): String = "CliqItModule"

  override fun getConstants(): MutableMap<String, Any> = hashMapOf(
    "LinkField" to hashMapOf(
      "destination" to LinkField.destination,
      "iosDestination" to LinkField.iosDestination,
      "androidDestination" to LinkField.androidDestination,
      "ogTitle" to LinkField.ogTitle,
      "ogDescription" to LinkField.ogDescription,
      "ogImage" to LinkField.ogImage,
      "ogUrl" to LinkField.ogUrl,
      "slug" to LinkField.slug,
      "webFallback" to LinkField.webFallback,
      "showInterstitial" to LinkField.showInterstitial,
      "isDeepLink" to LinkField.isDeepLink,
      "appleTeamId" to LinkField.appleTeamId,
      "iosBundleId" to LinkField.iosBundleId,
      "androidPackageName" to LinkField.androidPackageName,
      "resolvedPath" to LinkField.resolvedPath,
    ),
  )

  @ReactMethod
  fun addListener(eventName: String) {
    hasListeners = true
    flushPending()
  }

  @ReactMethod
  fun removeListeners(count: Int) {
    if (count > 0) hasListeners = false
  }

  @ReactMethod
  fun configure(apiKey: String) {
    CliqItSDK.init(reactContext)
    bindHandlersIfNeeded()
    CliqItSDK.configure(apiKey)
    CliqItSDK.notifyAlreadyReportedIfNeeded()
  }

  @ReactMethod
  fun handleUrl(url: String) {
    CliqItSDK.init(reactContext)
    CliqItSDK.handle(url)
  }

  private fun bindHandlersIfNeeded() {
    if (handlersBound) return
    handlersBound = true

    CliqItSDK.onLinkReceived { payload ->
      val query = Arguments.createMap()
      payload.query.forEach { (k, v) -> query.putString(k, v) }
      val components = Arguments.createArray()
      payload.pathComponents.forEach { components.pushString(it) }
      val map = Arguments.createMap().apply {
        putString("url", payload.url)
        putString("path", payload.path)
        putArray("pathComponents", components)
        putMap("query", query)
        putString("source", payload.source)
        putBoolean("isDeferred", payload.isDeferred)
        putString("status", payload.status)
        payload.matched?.let { putBoolean("matched", it) }
        putOpt(this, "tier", payload.tier)
        putOpt(this, "confidence", payload.confidence)
        payload.score?.let { putDouble("score", it) }
        putOpt(this, "slug", payload.slug)
        putOpt(this, "destinationPath", payload.destinationPath)
        putOpt(this, "error", payload.errorMessage)
        putBoolean("shouldNavigate", payload.shouldNavigate)
      }
      emitOrBuffer("CliqItLinkReceived", map) { pendingDeepLink = it }
    }

    CliqItSDK.onDirectLinkLookup { result ->
      val map = Arguments.createMap()
      when (result) {
        is LinkLookupResult.Resolved -> {
          val d = result.details
          map.putString("status", "resolved")
          putOpt(map, LinkField.destination, d.destination)
          putOpt(map, LinkField.iosDestination, d.iosDestination)
          putOpt(map, LinkField.androidDestination, d.androidDestination)
          putOpt(map, LinkField.ogTitle, d.ogTitle)
          putOpt(map, LinkField.ogDescription, d.ogDescription)
          putOpt(map, LinkField.ogImage, d.ogImage)
          putOpt(map, LinkField.ogUrl, d.ogUrl)
          putOpt(map, LinkField.slug, d.slug)
          putOpt(map, LinkField.webFallback, d.webFallback)
          putOpt(map, LinkField.showInterstitial, d.field(LinkField.showInterstitial))
          putOpt(map, LinkField.isDeepLink, d.field(LinkField.isDeepLink))
          putOpt(map, LinkField.appleTeamId, d.appleTeamId)
          putOpt(map, LinkField.iosBundleId, d.iosBundleId)
          putOpt(map, LinkField.androidPackageName, d.androidPackageName)
          putOpt(map, LinkField.resolvedPath, d.resolvedPath)
        }
        is LinkLookupResult.Failed -> {
          map.putString("status", "failed")
          map.putString("error", result.error)
        }
      }
      emitOrBuffer("CliqItLinkLookup", map) { pendingLinkLookup = it }
    }

    CliqItSDK.onVerify { outcome ->
      val map = Arguments.createMap()
      when (outcome) {
        is VerifyOutcome.Passed -> {
          map.putString("status", "ok")
          map.putBoolean("ok", true)
          putOpt(map, "appId", outcome.result.appId)
          putOpt(map, "appName", outcome.result.appName)
          map.putString("message", outcome.result.mismatchMessage)
          map.putMap("checks", checksMap(outcome.result.checks))
          map.putString("raw", outcome.result.rawJSON)
        }
        is VerifyOutcome.Mismatched -> {
          map.putString("status", "mismatch")
          map.putBoolean("ok", false)
          putOpt(map, "appId", outcome.result.appId)
          putOpt(map, "appName", outcome.result.appName)
          map.putString("message", outcome.result.mismatchMessage)
          map.putMap("checks", checksMap(outcome.result.checks))
          map.putString("raw", outcome.result.rawJSON)
        }
        is VerifyOutcome.Error -> {
          map.putString("status", "failed")
          map.putBoolean("ok", false)
          map.putString("error", outcome.message)
        }
      }
      emitOrBuffer("CliqItVerify", map) { pendingVerify = it }
    }
  }

  private fun checksMap(checks: Map<String, com.cliqit.sdk.VerifyCheck>): WritableMap {
    val out = Arguments.createMap()
    checks.forEach { (key, check) ->
      val item = Arguments.createMap()
      putOpt(item, "actual", check.actual)
      putOpt(item, "expected", check.expected)
      item.putBoolean("match", check.match)
      out.putMap(key, item)
    }
    return out
  }

  private fun emitOrBuffer(
    event: String,
    body: WritableMap,
    buffer: (WritableMap) -> Unit,
  ) {
    if (hasListeners) {
      sendEvent(event, body)
    } else {
      buffer(body)
    }
  }

  private fun flushPending() {
    pendingDeepLink?.let {
      pendingDeepLink = null
      sendEvent("CliqItLinkReceived", it)
    }
    pendingLinkLookup?.let {
      pendingLinkLookup = null
      sendEvent("CliqItLinkLookup", it)
    }
    pendingVerify?.let {
      pendingVerify = null
      sendEvent("CliqItVerify", it)
    }
  }

  private fun sendEvent(event: String, params: WritableMap) {
    if (!reactContext.hasActiveReactInstance()) return
    reactContext
      .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
      .emit(event, params)
  }

  private fun putOpt(map: WritableMap, key: String, value: String?) {
    if (value != null) map.putString(key, value) else map.putNull(key)
  }
}
